File: 5_Claude_Agent_SDK/typescript/src/3_workflows/1_sequential_workflow.ts
#!/usr/bin/env node
/**
 * Sequential Workflow Example - Agent Chain Pattern
 *
 * Demonstrates a multi-step sequential workflow where the output of one agent
 * becomes the input to the next agent. This is useful for:
 * - Data processing pipelines
 * - Multi-stage analysis
 * - Progressive refinement tasks
 *
 * Pattern: Agent A -> Agent B -> Agent C
 */

import { query, SDKAssistantMessage, SDKResultMessage } from '@anthropic-ai/claude-agent-sdk';

async function sequentialWorkflow() {
  /**
   * Sequential workflow: Research -> Analyze -> Summarize
   *
   * Step 1: Research agent gathers information about a topic
   * Step 2: Analysis agent examines the research findings
   * Step 3: Summary agent creates a concise final report
   */

  // Step 1: Research Agent
  console.log('='.repeat(60));
  console.log('STEP 1: Research Phase');
  console.log('='.repeat(60));

  const topic = "Benefits of TypeScript's type system for large-scale applications";

  let researchFindings = '';

  for await (const message of query({
    prompt: `Research the following topic and provide 3-5 key points: ${topic}`,
    options: {
      systemPrompt:
        'You are a research assistant. Gather key information about the topic and provide detailed findings.',
      model: 'sonnet',
    },
  })) {
    if (message.type === 'assistant') {
      const assistantMsg = message as SDKAssistantMessage;
      for (const block of assistantMsg.message.content) {
        if (block.type === 'text') {
          console.log(block.text);
          researchFindings += block.text;
        }
      }
    } else if (message.type === 'result') {
      const resultMsg = message as SDKResultMessage;
      if (resultMsg.total_cost_usd && resultMsg.total_cost_usd > 0) {
        console.log(`\nCost: $${resultMsg.total_cost_usd.toFixed(4)}`);
      }
    }
  }

  console.log('\n');

  // Step 2: Analysis Agent
  console.log('='.repeat(60));
  console.log('STEP 2: Analysis Phase');
  console.log('='.repeat(60));

  let analysisResults = '';

  for await (const message of query({
    prompt: `Analyze these research findings and provide critical insights:\n\n${researchFindings}`,
    options: {
      systemPrompt:
        'You are an analytical expert. Examine the research findings and identify strengths, weaknesses, and practical implications.',
      model: 'sonnet',
    },
  })) {
    if (message.type === 'assistant') {
      const assistantMsg = message as SDKAssistantMessage;
      for (const block of assistantMsg.message.content) {
        if (block.type === 'text') {
          console.log(block.text);
          analysisResults += block.text;
        }
      }
    } else if (message.type === 'result') {
      const resultMsg = message as SDKResultMessage;
      if (resultMsg.total_cost_usd && resultMsg.total_cost_usd > 0) {
        console.log(`\nCost: $${resultMsg.total_cost_usd.toFixed(4)}`);
      }
    }
  }

  console.log('\n');

  // Step 3: Summary Agent
  console.log('='.repeat(60));
  console.log('STEP 3: Summary Phase');
  console.log('='.repeat(60));

  console.log('\nFinal Summary:\n');

  for await (const message of query({
    prompt: `Create a concise executive summary based on this analysis:\n\n${analysisResults}`,
    options: {
      systemPrompt:
        'You are a summarization specialist. Create concise, actionable summaries suitable for executive review.',
      model: 'sonnet',
    },
  })) {
    if (message.type === 'assistant') {
      const assistantMsg = message as SDKAssistantMessage;
      for (const block of assistantMsg.message.content) {
        if (block.type === 'text') {
          console.log(block.text);
        }
      }
    } else if (message.type === 'result') {
      const resultMsg = message as SDKResultMessage;
      if (resultMsg.total_cost_usd && resultMsg.total_cost_usd > 0) {
        console.log(`\nCost: $${resultMsg.total_cost_usd.toFixed(4)}`);
      }
    }
  }

  console.log('\n');
  console.log('='.repeat(60));
  console.log('Sequential workflow completed!');
  console.log('='.repeat(60));
}

async function main() {
  await sequentialWorkflow();

  console.log('\n=== Sequential Workflow Summary ===');
  console.log('✓ Multi-step agent chain');
  console.log('✓ Output of one agent feeds into next');
  console.log('✓ Progressive refinement');
  console.log('✓ Linear data pipeline');
}

main().catch(console.error);

File: podklady/5_Claude_Agent_SDK/python/3_workflows/1_sequential_workflow.py
"""
Sequential Workflow Example - Agent Chain Pattern

Demonstrates a multi-step sequential workflow where the output of one agent
becomes the input to the next agent. Each agent runs in its own isolated
ClaudeSDKClient session. This is useful for:
- Data processing pipelines
- Multi-stage analysis
- Progressive refinement tasks

Pattern: Agent A -> Agent B -> Agent C
"""

import anyio

from claude_agent_sdk import (
    AssistantMessage,
    ClaudeAgentOptions,
    ClaudeSDKClient,
    ResultMessage,
    TextBlock,
)


async def run_agent(name: str, system_prompt: str, prompt: str) -> str:
    """Run a single agent in its own isolated session.

    Args:
        name: Display name for the agent
        system_prompt: System prompt defining the agent's role
        prompt: The task prompt

    Returns:
        The agent's text response
    """
    options = ClaudeAgentOptions(
        model="sonnet",
        system_prompt=system_prompt,
    )

    result = ""
    async with ClaudeSDKClient(options=options) as client:
        await client.query(prompt)
        async for msg in client.receive_response():
            if isinstance(msg, AssistantMessage):
                for block in msg.content:
                    if isinstance(block, TextBlock):
                        result += block.text
                        print(block.text, end="", flush=True)
            elif isinstance(msg, ResultMessage):
                if msg.total_cost_usd and msg.total_cost_usd > 0:
                    print(f"\nCost for {name}: ${msg.total_cost_usd:.4f}")

    return result


async def sequential_workflow():
    """
    Sequential workflow: Research -> Analyze -> Summarize

    Step 1: Research agent gathers information about a topic
    Step 2: Analysis agent examines the research findings
    Step 3: Summary agent creates a concise final report
    """

    topic = "Benefits of Python's asyncio for concurrent programming"

    # Step 1: Research Agent
    print("=" * 60)
    print("STEP 1: Research Phase")
    print("=" * 60)

    research_findings = await run_agent(
        name="Research Agent",
        system_prompt="You are a research assistant. Gather key information about the topic and provide detailed findings.",
        prompt=f"Research the following topic and provide 3-5 key points: {topic}",
    )

    print("\n")

    # Step 2: Analysis Agent
    print("=" * 60)
    print("STEP 2: Analysis Phase")
    print("=" * 60)

    analysis_results = await run_agent(
        name="Analysis Agent",
        system_prompt="You are an analytical expert. Examine the research findings and identify strengths, weaknesses, and practical implications.",
        prompt=f"Analyze these research findings and provide critical insights:\n\n{research_findings}",
    )

    print("\n")

    # Step 3: Summary Agent
    print("=" * 60)
    print("STEP 3: Summary Phase")
    print("=" * 60)

    print("\nFinal Summary:\n")
    summary = await run_agent(
        name="Summary Agent",
        system_prompt="You are a summarization specialist. Create concise, actionable summaries suitable for executive review.",
        prompt=f"Create a concise executive summary based on this analysis:\n\n{analysis_results}",
    )

    print("\n")
    print("=" * 60)
    print("Sequential workflow completed!")
    print("=" * 60)


if __name__ == "__main__":
    anyio.run(sequential_workflow)
